libbpf-tools/biostacks.c

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Wenbo Zhang
+//
+// Based on biostacks(8) from BPF-Perf-Tools-Book by Brendan Gregg.
+// 10-Aug-2020   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "biostacks.h"
+#include "biostacks.skel.h"
+#include "trace_helpers.h"
+
+static struct env {
+	char *disk;
+	int disk_len;
+	int duration;
+	bool milliseconds;
+	bool verbose;
+} env = {
+	.duration = -1,
+};
+
+const char *argp_program_version = "biostacks 0.1";
+const char *argp_program_bug_address = "<ethercflow@gmail.com>";
+const char argp_program_doc[] =
+"Tracing block I/O with init stacks.\n"
+"\n"
+"USAGE: biostacks [-h]\n"
+"\n"
+"EXAMPLES:\n"
+"    biostacks              # trace block I/O with init stacks.\n"
+"    biolatency -d sdc      # trace sdc only\n";
+
+static const struct argp_option opts[] = {
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{ "disk",  'd', "DISK",  0, "Trace this disk only" },
+	{ "milliseconds", 'm', NULL, 0, "Millisecond histogram" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'h':
+		argp_usage(state);
+		break;
+	case 'd':
+		env.disk = arg;
+		env.disk_len = strlen(arg) + 1;
+		if (env.disk_len > DISK_NAME_LEN) {
+			fprintf(stderr, "invaild disk name: too long\n");
+			argp_usage(state);
+		}
+		break;
+	case 'm':
+		env.milliseconds = true;
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++) {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		env.duration = strtoll(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "invalid delay (in us): %s\n", arg);
+			argp_usage(state);
+		}
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+static int print_map(struct ksyms *ksyms, int fd)
+{
+	char *units = env.milliseconds ? "msecs" : "usecs";
+	struct rqinfo lookup_key = {}, next_key;
+	const struct ksym *ksym;
+	int num_stack, i, err;
+	struct hist hist;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		printf("%-14.14s %-6d %-7s\n",
+			next_key.comm, next_key.pid, next_key.disk);
+		num_stack = next_key.kern_stack_size /
+			sizeof(next_key.kern_stack[0]);
+		for (i = 0; i < num_stack; i++) {
+			ksym = ksyms__map_addr(ksyms, next_key.kern_stack[i]);
```

Comment:
ksym could be NULL, you need to handle that

Reply:
Oh, sorry for forgot this, I'll fix this then.